Add description for BAT board on documentation (#313)
icub-tech-iit/documentation: docs/battery_boards/bat_board/bat_protocol.md

# BAT board Protocol

## Introduction

> This page describes the `BAT` board and its protocol. Basically, this board is responsible for the management of the data and signals coming from the battery pack in `R1`, `iCub` and `ergoCub` robots. Moreover, it should be noted that in the latter robots mentioned, it is always coupled with the `BMS` baord, which is responsible for a more fine tuning and managing of the battery pack status.

Reviewer: ```suggestion
This page describes the `BAT` board and its protocol. Basically, this board is responsible for the management of the data and signals provided by the battery pack in `R1`, `iCub`, and `ergoCub` robots. Moreover, it should be noted that in these robots it is always coupled with the `BMS` baord, which is responsible for fine-tuning and managing the battery pack status.
```

> In general, as shown in the figures below (which actually show the connection between the BAT and EMS board in the base of `R1/SN003`), the `BAT` board can be connected to an EMS board through the specific `CAN` connector, which is responsible for reading the CAN frames coming from the `BAT`.

Reviewer: ```suggestion
In general, as shown in the figures below (illustrating the connection between the BAT and EMS boards in the base of `R1/SN003`), the `BAT` can be connected to an EMS board through the `CAN` connector devoted to receiving the CAN frames sent out by the `BAT`.
```

<center>
<table><tr>
<td><img width="500" height="500" src="../assets/bat-2-ems-R1-connection-01.png"/></td>
<td><img width="500" height="500" src="../assets/bat-2-ems-R1-connection-02.png"/></td>
</tr></table>
</center>

> Specifically, at the current status of the developments (July 2023), the `BAT` board can send the following data:
> 
> - Battery Pack Voltage in Volt
> - Battery Pack Charge in $\%$
> - Battery Pack Status (details about it will follow)

Reviewer: ```suggestion
Specifically, at the current status of the development (July 2023), the `BAT` board can send the following data:

- Battery Pack Voltage in Volt.
- Battery Pack Charge in $\%$.
- Battery Pack Status (detail will follow).
```

### Communication characteristics

> The pieces of information detailed above are sent by the `BAT` board with a FIFO cycle of 1ms. Then, the `EMS` board handles the CAN frames sent by the `BAT`, parses them and finally forwards them to the higher level of the `yarprobotinterface` at the specific port defined in the configuration files with a frequence o 1Hz.

Reviewer: ```suggestion
The pieces of information detailed above are sent by the `BAT` board with a FIFO cycle of 1 ms. Then, the `EMS` board handles the CAN frames sent by the `BAT`, parses them, and finally forwards them to the higher level of the `yarprobotinterface` at the specific port defined in the configuration files with a frequency of 1 Hz.
```

> Furthermore, those info are also sent directly from the `BAT` to the display attached to the robot, every 10ms.

Reviewer: ```suggestion
Furthermore, that info is also sent directly from the `BAT` to the display attached to the robot, every 10 ms.
```

Remember that info is uncountable.

    Position | BIT[7] | BIT[6] | BIT[5] | BIT[4] | BIT[3] | BIT[2] | BIT[1] | BIT[0] |
    |:---|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
    Value | V12board | V12board_F | V12motor | V12motor_F | HSM | HSM_PG | HSM_F | HSM_broken |
    Description | 12V DCDC board regulator | 12V DCDC board regulator OVERCURRENT fault | 12V DCDC motor regulator | 12V DCDC motor regulator OVERCURRENT fault | Hot Swap Manager | Hot Swap Manager POWER GOOD | Hot Swap Manager OVERCURRENT/OVERVOLTAGE fault | Hot Swap Manager MOSFETs damaged |
    Possible status | ON(1)/OFF(0) | OC/NORMAL | ON/OFF | OC/NORMAL| ON/OFF | HSM output voltage stable after transient/HSM output voltage not guaranteed | OC-OV/NORMAL | HSM MOSFETs probably burned/NORMAL |

- DCDC_status_B:

    Position | BIT[1] | BIT[0] |
    |:---|:---:|:---:|
    Value | PB1_restart | PB2_restart |
    Description | Restart phase of the push button 1 | Restart phase of the push button 2 |
    Possible status | Start-up phase(1)/stable operation(0) | Start-up phase/stable operation |

- Final status shown at the port is equal to: 

    `(DCDC_status_B << 8 ) | DCDC_status_A`

### Data displayed on the yarp port

> The user gets the data from a specific YARP port defined in the configuration file as follows:

Reviewer: ```suggestion
The user gets the data from a specific YARP port defined in the configuration file. Here's the format:
```

> It should be noted that as things now the user will see the data at the specific yarp port defined in the configuration file as follows:
> 
> - Voltage will be displayed in the Volts
> - Battery charge will be displayed in $\%$
> - Status will be displayed as an integer of 16 bits (where we fill only the first 10), which can be interpreted following the tables above

Reviewer: ```suggestion
The user gets the data from a specific YARP port defined in the configuration file as follows:

- Voltage is displayed in the Volts.
- Battery charge is displayed in $\%$.
- Status is displayed as a 16-bit integer (only the first 10 bits are valid), whose mapping adheres to the tables above.
```

> Moreover, when the boards start a **DEBUG** message with the initial values of the status (converted to the description strings) is send to the `yarprobotinterface`. 

Reviewer: ```suggestion
Moreover, at start-up, a **DEBUG** message with the initial values of the status (converted to the description strings) is sent to `yarprobotinterface`. 
```

> Then, each time the status changes a **DEBUG** message is send to the `yarprobotinterface` that shows only, as description strings, the value of the bits changed with respect to the previous status.

Reviewer: ```suggestion
Then, each time the status changes, a **DEBUG** message is sent to `yarprobotinterface`, which in turn prints out a description only about the values of those bits that have switched.
```
